docs/ADR-8-Use-RIO-in-cardano‐cli.md

```diff
@@ -0,0 +1,129 @@
+
+The vast majority of these errors are not used to recover, they are propagated and reported to the user.
+```
+main :: IO ()
+main = toplevelExceptionHandler $ do
+  envCli <- getEnvCli
+  co <- Opt.customExecParser pref (opts envCli)
+  orDie (docToText . renderClientCommandError) $ runClientCommand co
+  
+...
+
+runClientCommand :: ClientCommand -> ExceptT ClientCommandErrors IO ()
+runClientCommand = \case
+  AnyEraCommand cmds ->
+    firstExceptT (CmdError (renderAnyEraCommand cmds)) $ runAnyEraCommand cmds
+  AddressCommand cmds ->
+    firstExceptT AddressCmdError $ runAddressCmds cmds
+  NodeCommands cmds ->
+    runNodeCmds cmds
+      & firstExceptT NodeCmdError
+  ByronCommand cmds ->
+    firstExceptT ByronClientError $ runByronClientCommand cmds
+  CompatibleCommands cmd ->
+    firstExceptT (BackwardCompatibleError (renderAnyCompatibleCommand cmd)) $
+      runAnyCompatibleCommand cmd
+...
+```
+- As a result we have a lot of errors wrapped in errors which makes the code unwieldly and difficult to compose with other code blocks. See image below of incidences of poor composability where we use `firstExceptT` (and sometimes `first`) to wrap errors in other errors.
+
+![image](https://github.com/user-attachments/assets/6a50e8f2-9140-4b7e-afa5-ff778d696742)
+ 
+- The ExceptT IO is a known anti-pattern for these reasons and others as per:
+  - https://github.com/haskell-effectful/effectful/blob/master/transformers.md
+  - https://tech.fpcomplete.com/blog/2016/11/exceptions-best-practices-haskell/
+
+
+# Proposed Solution
+
+Use the RIO monad which is based on the [ReaderT design pattern](https://tech.fpcomplete.com/blog/2017/06/readert-design-pattern/).
+
+### Pros 
+I opt for the `RIO env a` monad instead of `ReaderT e m a` because: 
+- We get additional [logging functionality](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#g:5)
+- Explicit environment dependencies e.g `logError :: HasLogFunc env => Text -> RIO env ()`
+- Better composability i.e no more errors that wrap errors (see above).
+
+### Cons
+- `RIO` is hardcoded to IO so we cannot add additional transformer layers e.g `RIO Env (StateT Int IO) a`
+- Implicit error flow. Errors are thrown via `IO` exceptions.
+
+### Exception handling
+
+#### Exception type
+```haskell
+data CustomCliException where
+  CustomCliException
+    :: (Show error, Typeable error, Error error, HasCallStack)
+    => error -> CustomCliException
+
+deriving instance Show CustomCliException
+
+instance Exception CustomCliException where
+  displayException (CustomCliException e) =
+    unlines
+      [ show (prettyError e)
+      , prettyCallStack callStack
+      ]
+```
+
+#### Top level handler 
+```haskell
+executeRio :: RIO () () -> IO ()
+executeRio r = do
+  runRIO () r
+    `catch` ( \e ->
+                case fromException e of
+                  Just custom@(CustomCliException{}) -> putStrLn $ displayException custom
```

Comment:
1. [ADR 7](https://github.com/input-output-hk/cardano-node-wiki/wiki/ADR-7-CLI-Output-Presentation) says we should log to stderr
2. I'm not quite seeing the point of differentiation between two cases: `CustomCliException` and the other one. You're doing `displayException` in either case.

Reply:
>This ensures we capture all exceptions.

Catching `SomeException` would do the same here.


>We handle CustomCliException explicitly for structured, user-facing errors, while the Nothing case acts as a safety net for unexpected exceptions.

Again, this separation is not shown anywhere in the code. A next person reading this ADR will have no clue what is the role of `CustomCliException`.

Reply:
> Catching SomeException would do the same here.

Are you saying to explicitly pattern match on this? I don't understand your point here.

Reply:
>Are you saying to explicitly pattern match on this? I don't understand your point here.
Yes, `SomeException` is a catch all for all exceptions (asynchronous as well).

Reply:
How is this different that the `Nothing` case? It's going to handle all other exceptions.

Reply:
Exactly, it is not - so what's specific about `CustomCliException` to pattern match it in the handler?

Reply:
https://github.com/input-output-hk/cardano-node-wiki/pull/60#discussion_r1943309730